src/btree/bt_cursor.c

```diff
@@ -1360,9 +1365,19 @@ __cursor_chain_exceeded(WT_CURSOR_BTREE *cbt)
 	    page->modify != NULL && page->modify->mod_row_update != NULL)
 		upd = page->modify->mod_row_update[cbt->slot];
 
-	for (i = 0; upd != NULL; ++i, upd = upd->next) {
+	for (i = 0, upd_size = 0; upd != NULL; ++i, upd = upd->next) {
 		if (WT_UPDATE_DATA_VALUE(upd))
 			return (false);
+		upd_size += WT_UPDATE_MEMSIZE(upd);
+		if (!__wt_txn_upd_visible_all(session, upd))
+			continue;
+		/*
+		 * When history has to be maintained, creating extra copies
+		 * of large documents multiplies cache pressure, so avoid it,
+		 * by growing the modify chain.
+		 */
+		if (upd_size >= WT_MODIFY_MEM_FACTOR * cursor->value.size)
+			return (true);
```

Comment:
I'm currently testing the performance with this check moved above the `__wt_txn_upd_visible_all` check.  In other words, when updates are a large proportion of the underlying record size, don't keep too many around, even if not all of them are visible.  The change still allows @bvpvamsikrishna's test to complete, but seems to have less impact on performance of automated tests.

I'll report back on the ticket when I have more solid data.